src/test/resources/normalizer/weirdToNormalizedUrls.csv

```diff
@@ -124,7 +124,7 @@ http://foo.com/aa/bb//foo.html, http://foo.com/aa/bb/foo.html
 http://foo.com//aa//bb//foo.html, http://foo.com/aa/bb/foo.html
 http://foo.com////aa////bb//foo.html, http://foo.com/aa/bb/foo.html
 http://foo.com////aa////bb////foo.html, http://foo.com/aa/bb/foo.html
-http://foo.com/aa?referer=http://bar.com, http://foo.com/aa?referer=http://bar.com
```

Comment:
I was working on this change and found the following existing test cases which seem to be inconsistent:

1. `http://x.com/show?http%3A%2F%2Fx.com%2Fb` => `http://x.com/show?http%3A%2F%2Fx.com%2Fb`
2. `http://foo.com/aa?referer=http://bar.com` => `http://foo.com/aa?referer=http://bar.com`

Even though the first case is malformed with an incomplete query name-value pair,  the string `http%3A%2F%2Fx.com%2Fb` should likely be interpreted as a query name. If we are not encoding ':' and '/', then shouldn't the correct normalization be: 
`http://x.com/show?http%3A%2F%2Fx.com%2Fb` => `http://x.com/show?http://x.com/b`

Reply:
Hi @aecio, the query part of a URL is not necessarily filled from a form. This means it can contain any arbitrary query string which isn't a key-value structure and with a encoding not (fully) consistent with `application/x-www-form-urlencoded`. Just a few examples:
- http://etab.ac-poitiers.fr/lepmo/spip.php?article247
- https://norma-jackson-fine-art.com/gallery/index.php?/category/52
- https://cmsimplexh.webdesign-keil.de/?Fachthemen/Ein_CMSimple_XH_Template_erstellen/Idee_%2F_Planung - here `/` and `%2F` have a different semantic replacing them cause a [page not found](https://cmsimplexh.webdesign-keil.de/?Fachthemen%2FEin_CMSimple_XH_Template_erstellen%2FIdee_%2F_Planung)

Of course, most often the [query string](https://en.wikipedia.org/wiki/Query_string) stems from a HTML form. However, it's also important not to break valid URLs. That's what I mean with finding the "balance between strict normalization (eg. to avoid duplicates) and the preservation of potential semantically different URLs".

Reply:
I understand we should not change the URL when there is potential for breaking it. Your last example shows an interesting edge case where using a standard decoding could create problems. What is still not clear to me is what is the minimal set of normalizations that are safe to apply to the query data. 
- Should we not change the data query data at all?
- When should we sort name-value pairs? When doing so, should we keep data encoding intact?
- Is there any case we want to change the data encoding (maybe only when there are illegal characters)?

Reply:
> the minimal set of normalizations that are safe to apply to the query data
> Is there any case we want to change the data encoding...

See [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.4) and the [whatwg URL spec](https://url.spec.whatwg.org/#query-percent-encode-set) mentioned above. There's an interesting discussion in whatwg/url#369 and a nice JavaScript function to test how browsers normalize URLs.

> Should we not change the data query data at all?

What about making this optional (maybe on by default)? Your PR already adds a constructor to configure the removal of parameters by name. Any good ideas how to make the URL normalization configurable?

Reply:
To make the URL normalization more configurable, we can change the constructor to a builder object that contains all configurable options. This would allow adding many options without having a complex constructor and while still keeping a simple fluent API with reasonable defaults options.

I think that for this case, the options could be the following?

    urlEncodeQueryParameters = NEVER | ALWAYS | KEY_VALUE_PAIRS_ONLY

If so, what would be the desired default? `KEY_VALUE_PAIRS_ONLY`?